### PR TITLE
Added incremental building of ensembles and fixed adaboost

### PR DESCRIPTION
**Pull request by nitbix**
Adaboost implementation had a bug which resulted in the D vector of dataset weights being incorrect and in some extreme cases leading to non-termination and infinity errors.

Ensembles now can have "steps" where they have a member added on to the existing ensemble. This only works for some ensemble techniques that can be expanded this way.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,8 +1,16 @@
...
+<<<<<<< HEAD
+*.class
+*.o
+*.so
+!libaparapi_x86_64.so
+cache.bin
+=======
 .DS_Store
 .gradle
+>>>>>>> refs/remotes/choose_remote_name/master
```
Comment (janisz) [inline .gitignore]: Looks like unresolved conflict

File: src/main/java/org/encog/util/normalize/output/OutputFieldRangeMapped.java [hunk 2]
```diff
@@ -101,10 +101,13 @@ public OutputFieldRangeMapped(InputField f) {
...
 	public double calculate(final int subfield) {
-		return ((this.field.getCurrentValue() - this.field.getMin()) 
-				/ (this.field
-				.getMax() - this.field.getMin()))
-				* (this.high - this.low) + this.low;
+		if (this.field.getMax() == this.field.getMin()) {
+			return 0;
+		} else {
+			return ((this.field.getCurrentValue() - this.field.getMin()) 
+					/ (this.field.getMax() - this.field.getMin()))
```
Comment (janisz) [inline src/main/java/org/encog/util/normalize/output/OutputFieldRangeMapped.java]: `this.field` should have method that return normalized value to make code more redable

**Comment by jeffheaton** [comment]
I finally merged this in, very sorry about the long delay on this one.  I've actually gotten into some of the technology you've added, so this is very interesting for me to go through this.  I am going to experiment with my existing Encog code and make sure everything works as I expect, and will make any changes as needed.

**Comment by nitbix** [comment]
Thanks, do let me know if anything is unclear or looks wrong!

On 7 July 2015 at 18:29, Jeff Heaton <notifications@github.com> wrote:

> I finally merged this in, very sorry about the long delay on this one.
> I've actually gotten into some of the technology you've added, so this is
> very interesting for me to go through this. I am going to experiment with
> my existing Encog code and make sure everything works as I expect, and will
> make any changes as needed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/encog/encog-java-core/pull/138#issuecomment-119276410>
> .
>



-- 
  Alan

**Comment by jeffheaton** [comment]
Just one small thing I noticed, the FlatNetwork does have a default constructor that is used sometimes with serialization.  It does not call init, so the dropout percentage variable was not getting set and causing NullPointerException's later on.  The following very simple fix resolves that:  

https://github.com/encog/encog-java-core/commit/03613385836192b5ef91b1dd5f7037d852df4fc7

Otherwise all of the previous code seems to be functioning as I would expect!

Thanks again!  I am sure I will have more questions as I go through it.

**Comment by jeffheaton** [comment]
FYI, the paper with my citation to your thesis for this has been published.  http://jmlr.org/papers/volume16/heaton15a/heaton15a.pdf

**Pull request merged**